File: python/circular_buffer.py
class BufferFullException(BufferError):
    """Exception raised when CircularBuffer is full.

    message: explanation of the error.

    """

    def __init__(self, message):
        self.message = message


class BufferEmptyException(BufferError):
    """Exception raised when CircularBuffer is empty.

    message: explanation of the error.

    """

    def __init__(self, message):
        self.message = message


class CircularBuffer:

    def __init__(self, capacity):
        self._buf = [None] * capacity
        self._front = 0
        self._rear = 0
        self._len = 0
        self._capacity = capacity

    def read(self):
        if self.is_empty():
            raise BufferEmptyException('Circular buffer is empty')
        val = self._buf[self._front]
        self._front = self.next(self._front)
        self._len -= 1
        return val

    def write(self, data):
        if self.is_full():
            raise BufferFullException('Circular buffer is full')
        self._buf[self._rear] = data
        self._rear = self.next(self._rear)
        self._len += 1

    def overwrite(self, data):
        if self.is_full():
            self._buf[self._front] = data
            self._front = self.next(self._front)
        else:
            self.write(data)

    def clear(self):
        self._buf = [None] * self._capacity
        self._front = 0
        self._rear = 0
        self._len = 0

    def next(self, curr):
        return (curr + 1) % self._capacity

    def is_empty(self):
        return self._len == 0

    def is_full(self):
        return self._len == self._capacity

File: lua/circular-buffer.lua
local CircularBuffer = {}
CircularBuffer.__index = CircularBuffer

function CircularBuffer:new(cap)
    local t = {
        front = 1,
        rear = 1,
        len = 0,
        cap = cap,
        buf = {},
    }
    return setmetatable(t, CircularBuffer)
end

local function bump(n, len)
    return n % len + 1
end

function CircularBuffer:read()
    assert(self.len > 0 and self.len <= self.cap, "buffer is empty")
    local r = self.buf[self.front]
    self.front = bump(self.front, self.cap)
    self.len = self.len - 1
    return r
end

function CircularBuffer:write(n)
    if n == nil then
        return
    end
    assert(self.len < self.cap, "buffer is full")
    self.buf[self.rear] = n
    self.rear = bump(self.rear, self.cap)
    self.len = self.len + 1
end

function CircularBuffer:clear(n)
    self.buf = {}
    self.front = 1
    self.rear = 1
    self.len = 0
end

function CircularBuffer:forceWrite(n)
    if self.len < self.cap then
        return self:write(n)
    end
    self.buf[self.front] = n
    self.front = bump(self.front, self.cap)
end

return CircularBuffer

